### PR TITLE
refactor(Forms): EO-1096 refactor select forms

### PR DESCRIPTION
**Pull request by manojdas2212**
JIRA: https://jira.rax.io/browse/EO-1096

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] Design LGTM
- [x] QE LGTM

Before
------
screen shot 2016-08-25 at 6 40 02 pm

After
------
screen shot 2016-08-25 at 6 40 48 pm

Full Page
------
screen shot 2016-08-30 at 7 15 01 pm
screen shot 2016-08-30 at 7 15 14 pm

API Doc
------
screen shot 2016-08-25 at 6 41 49 pm

Search for "All Modules"
------
Search for rxSelect
------
screen shot 2016-08-25 at 6 42 23 pm

Search for select
------
screen shot 2016-08-25 at 6 42 36 pm

Search for dropdown
------
screen shot 2016-08-25 at 6 42 47 pm

Search for form
------
screen shot 2016-08-25 at 6 43 00 pm

Search for input
------
screen shot 2016-08-25 at 6 43 12 pm

### CODE REVIEW

**Comment by manojdas2212** [comment]
@CITguy I am not sure why midways are failing in this file utils/rx-page-objects/test/rxCheckbox.midway.js not even modify anything in that

screen shot 2016-08-25 at 9 37 29 pm

File: src/elements/Forms/docs/examples/rxSelect.simple.html [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<div ng-controller="rxSelectCtrl">
+<div ng-controller="rxSelectSimpleCtrl">
   <div class="demoContainer">
     <h3>Examples</h3>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.simple.html]: 1. `Select States` should not be in an example, but in the `Forms/docs/forms.docs.html` file
2. `Plain HTML Select Boxes (for comparison)` should not be in an example either. It should be contained inside a `<rx-debug>` element within `Forms/docs/forms.docs.html`.
3. `Plain HTML Select Boxes (for comparison)` has duplicate `id` values causing confusion for the selectors in the `rxCheckbox.midway.js` tests. Please update to using unique `id` values. I'd recommend `plainSelDisabled` and `plainSelNormal` for the appropriate `id` values.

File: src/elements/Forms/docs/examples/rxSelect.simple.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 angular.module('demoApp')
-.controller('rxSelectCtrl', function ($scope) {
+.controller('rxSelectSimpleCtrl', function ($scope) {
     $scope.radCreateDestroy = 'destroyed';
 
     $scope.validEnabled = 3;
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.simple.js]: This file should be broken up according to the example split mentioned above.

1. The `valid*`, `invalid*`, and `htmlSelectAlternativeValue` scope variables should remain in a demo-only controller located at `Forms/docs/rxSelect.js` (see `Forms.docs/rxCheckbox.js` as an example)

File: src/elements/Forms/docs/forms.docs.html [hunk 3]
```diff
@@ -132,17 +132,22 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
...
-<h3><rx-permalink>Selects</rx-permalink></h3>
+<h3>
+  <rx-permalink>Selects</rx-permalink>
+  <a class="button xs inline" href="ngdocs/index.html#/api/elements.directive:rxSelect/">View API</a>
+</h3>
+
 <ul class="list">
   <li>For single item selection, use the
-    <a href="#/components/rxSelect">rxSelect</a> component.
+    <a href="#/elements/Forms#Select">rxSelect</a> component.
```
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: Link is unnecessary. Reader is already viewing the section linked.

Please update to use plain text...

```
<li>For single item selection, use the rxSelect directive.</li>
```

File: src/elements/Forms/docs/forms.docs.html [hunk 4]
```diff
@@ -330,6 +421,48 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
            id="plainChkRemoveable"
            ng-if="!plainChkIsRemoved" />
   </p>
+
+  <h3>Plain HTML Select Boxes (for comparison)</h3>
```
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: These select boxes are in the wrong location. Please move them up with the rest of the rxSelect documentation.

File: src/elements/Forms/docs/forms.docs.html [hunk 5]
```diff
@@ -330,6 +421,48 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
...
+    <select id="plainHtmlNormal" ng-required="true">
+      <option value="">Plain HTML Select Option</option>
+      <option value="first">First Option</option>
+      <option value="second">Second Option</option>
+      <option value="third">Third Option</option>
+      <option value="fourth">Fourth Option</option>
+    </select>
+  </p>
+  <p>
+    <select id="plainHtmlDisabled" disabled>
```
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: The `id` attribute is not unique on the page. Please change.

I'd recommend changing to `plainSelDisabled`.

File: src/elements/Forms/docs/forms.docs.html [hunk 6]
```diff
@@ -330,6 +421,48 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
...
+      <option value="">Disabled HTML Select Option</option>
+      <option value="first">First Option</option>
+      <option value="second">Second Option</option>
+      <option value="third">Third Option</option>
+      <option value="fourth">Fourth Option</option>
+    </select>
+  </p>
+  <p>
+    <select id="plainHtmlSecondSelected"
```
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: The `id` attribute is not unique on the page. Please change.

I'd recommend changing to `plainSelSecondSelected`.

File: utils/rx-page-objects/test/rxCheckbox.midway.js [hunk 7]
```diff
@@ -91,14 +91,14 @@ describe('rxCheckbox', function () {
 
     describe('plain HTML checkboxes', function () {
         describe('Valid Enabled Unchecked', encore.exercise.rxCheckbox({
-            instance: encore.rxCheckbox.initialize($('#plainHtmlNormal')),
+            instance: encore.rxCheckbox.initialize($('#plainSelNormal')),
```
Comment (CITguy) [inline utils/rx-page-objects/test/rxCheckbox.midway.js]: Please revert these changes. The plain HTML checkboxes do not need modified.

File: utils/rx-page-objects/test/rxCheckbox.midway.js [hunk 8]
```diff
@@ -91,14 +91,14 @@ describe('rxCheckbox', function () {
...
             disabled: false,
             selected: false,
             valid: false
         }));
 
         describe('Valid Disabled Unchecked', encore.exercise.rxCheckbox({
-            instance: encore.rxCheckbox.initialize($('#plainHtmlDisabled')),
+            instance: encore.rxCheckbox.initialize($('#plainSelDisabled')),
```
Comment (CITguy) [inline utils/rx-page-objects/test/rxCheckbox.midway.js]: Please revert change.

File: src/elements/Forms/docs/forms.docs.html [hunk 9]
```diff
@@ -132,18 +132,109 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
...
     the <a href="#/components/rxMultiSelect">rxMultiSelect</a>
     component.
   </li>
 </ul>
 
+<h3>With Validation</h3>
+<rx-example name="rxSelectValidation.simple"></rx-example>
+<h3>Show/Hide Select</h3>
+<rx-example name="rxSelectshowHide.simple"></rx-example>
+<h3>Destroy Select</h3>
```
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: Please add blank lines above each heading, for readability.

File: src/elements/Forms/docs/examples/rxSelectValidation.simple.html [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  <select rx-select
+    id="selBaconPrep"
+    ng-model="baconPrep"
+    ng-required="true">
+    <option value="">I do not like bacon</option>
+    <option value="thin">Thin (light and crispy)</option>
+    <option value="medium">Medium (perfect balance of flavor)</option>
+    <option value="thick">Thick (borderline jerky)</option>
+    <option value="crumbled">Crumbled (great on salads)</option>
+  </select>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelectValidation.simple.html]: Please update example to use proper `rxForm` markup.

```html
<form rx-form>
  <rx-form-section controlled-width>
    <rx-field>
      ...
    </rx-field>
  </rx-form-section>
</form>
```

File: src/elements/Forms/docs/examples/rxSelectshowHide.simple.html [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<p>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelectshowHide.simple.html]: Please update file name to begin with `rxSelect.*`

In this case, I'd recommend `rxSelect.showHide.html`

File: src/elements/Forms/docs/examples/rxSelectValidation.simple.html [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<p>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelectValidation.simple.html]: Please update file name to begin with `rxSelect.*`

In this case, I'd recommend `rxSelect.simple.html`

File: src/elements/Forms/docs/examples/rxSelectDestroy.simple.html [hunk 13]
```diff
@@ -0,0 +1,29 @@
+<div ng-controller="rxSelectSimpleCtrl">
+  <div class="demoContainer">
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelectDestroy.simple.html]: 1. Change file name to `rxSelect.destroy.html`
1. Change controller name to `rxSelectDestroyCtrl`
1. Update markup to use proper `rxForm` directive markup.
1. The `.demoContainer` CSS class is no longer necessary. Use `<rx-form-section controlled-width>` instead.

```html
<form rx-form ng-controller="rxSelectDestroyCtrl">
  <rx-form-section controlled-width>
    ...
  </rx-form-section>
</form>
```

**Comment by CITguy** [comment]
`src/elements/Forms/styles/rxSelect.demo.less` is not necessary. Please remove.

File: src/elements/Forms/docs/examples/rxSelectshowHide.simple.html [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+  <input rx-checkbox id="chkShow" ng-model="isShown" />
+  <label for="chkShow">Show?</label><br />
+  <span ng-show="isShown">
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelectshowHide.simple.html]: Please update example to use proper `rxForm` markup.

```html
<form rx-form>
  <rx-form-section controlled-width>
    ...
  </rx-form-section>
</form>
```

File: src/elements/Forms/docs/forms.docs.html (hunk 4 above)
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: Additionally, you'll need to add `ng-controller="rxSelectCtrl"` to the `<rx-debug>` when you move this. This will allow the "Non Default Starting Option" value to be selected (as expected from the original midway tests).

```html
<rx-debug ng-controller="rxSelectCtrl">
  ...
</rx-debug>
```

File: utils/rx-page-objects/test/rxSelect.midway.js [hunk 15]
```diff
@@ -72,7 +72,7 @@ describe('rxSelect', function () {
             instance: encore.rxSelect.initialize($('#plainHtmlSecondSelected')),
             disabled: false,
             valid: true,
-            selectedText: 'Non Default Starting Option'
+            selectedText: 'Starts on Second Option'
```
Comment (CITguy) [inline utils/rx-page-objects/test/rxSelect.midway.js]: Please revert. Changes to be made with regard to the implementation of [simple HTML selects](https://github.com/rackerlabs/encore-ui/pull/1784/files#r76665241) should allow this to pass without modification.

**Comment by manojdas2212** [comment]
@CITguy I have addressed the review comments please take a look

**Comment by manojdas2212** [comment]
@100stacks @Droogans @ssalinas24 Please take a look this PR

File: utils/rx-page-objects/test/rxRadio.midway.js [hunk 16]
```diff
@@ -255,7 +255,7 @@ describe('rxRadio', function () {
 
             it('target radio should be present', function () {
                 radTargetCreated = rxRadioPage.initialize($('#radTargetCreated'));
-                expect(radTargetCreated.isPresent()).to.eventually.be.true;
+                expect(radTargetCreated.isPresent()).to.eventually.be.false;
             });
```
Comment (Droogans) [inline utils/rx-page-objects/test/rxRadio.midway.js]: Why would you change a test to make the premise of it incorrect? This isn't reflecting what's happening on the page, you've flipped a boolean to get the tests passing.

Comment (Droogans) [inline utils/rx-page-objects/test/rxRadio.midway.js]: Looks like this was supposed to test the creation/deletion of a *select box*, not a radio input.

If you change this radio element to a select box, as well as updating the id to `'selTargetCreated'`, this will work again.

**Comment by manojdas2212** [comment]
@Droogans made changes please take a look

File: src/elements/Forms/docs/examples/rxSelect.simple.html [hunk 17]
```diff
@@ -0,0 +1,26 @@
+<form rx-form>
+  <rx-form-section controlled-width>
+    <rx-field>
+      <rx-field-name></rx-field-name>
+      <rx-field-content>
+        <rx-input>
+          <p>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.simple.html]: `<p>` is unnecessary. Please remove.

File: src/elements/Forms/docs/examples/rxSelect.simple.html [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+            <strong>How do you like your bacon?</strong>
+            <small ng-if="baconPrep">({{baconPrep}})</small>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.simple.html]: Move content in lines 8-9 to the `<rx-field-name>` above and remove the `<strong>` tag.

```html
<rx-field-name>
  How do you like your bacon?
  <small ng-if="baconPrep">({{baconPrep}})</small>
</rx-field-name>
```

File: src/elements/Forms/docs/examples/rxSelect.simple.html [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+            <br />
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.simple.html]: `<br />` is unnecessary. Please remove.

File: src/elements/Forms/docs/examples/rxSelect.simple.html (hunk 18 above)
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.simple.html]: before | after
----- | -----
screen shot 2016-09-01 at 1 33 03 pm | screen shot 2016-09-01 at 1 33 55 pm

File: src/elements/Forms/docs/examples/rxSelect.showHide.html [hunk 20]
```diff
@@ -0,0 +1,20 @@
+<form rx-form>
+  <rx-form-section controlled-width>
+    <rx-field>
+      <rx-field-name></rx-field-name>
+      <rx-field-content>
+        <rx-input>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.showHide.html]: Please update lines 6-16 to the following:

```html
<rx-input>
  <rx-prefix>
    <input rx-checkbox id="chkShow" ng-model="isShown" />
    <label for="chkShow">Show?</label>
  </rx-prefix>
</rx-input>
<rx-input ng-show="isShown">
  <select rx-select id="selTargetShow">
    <option>I'm visible!</option>
  </select>
</rx-input>
```

This will allow the rxSelect to obtain the proper styling.

### before
screen shot 2016-09-01 at 1 47 50 pm


### after
screen shot 2016-09-01 at 1 46 13 pm

File: src/elements/Forms/docs/examples/rxSelect.showHide.html [hunk 21]
```diff
@@ -0,0 +1,20 @@
+<form rx-form>
+  <rx-form-section controlled-width>
+    <rx-field>
+      <rx-field-name></rx-field-name>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.showHide.html]: `<rx-field-name>` is unnecessary. Please remove.

File: src/elements/Forms/docs/examples/rxSelect.destroy.html [hunk 22]
```diff
@@ -0,0 +1,38 @@
+<form rx-form ng-controller="rxSelectDestroyCtrl">
+  <rx-form-section controlled-width>
+    <rx-field>
+      <rx-field-name></rx-field-name>
+      <rx-field-content>
+        <rx-input>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.destroy.html]: To adhere to the vertical radio option pattern and so that rxSelect gets proper styling, modify lines 6-34 to the following:

```html
<rx-input>
  <input 
    rx-radio
    id="radDestroyed"
    value="destroyed"
    ng-model="radCreateDestroy" />
  <label for="radDestroyed">Destroyed</label>
</rx-input>
<rx-input>
  <input 
    rx-radio
    id="radCreated"
    value="created"
    ng-model="radCreateDestroy" />
  <label for="radCreated">Created</label>
</rx-input>
<p>
  The following select is <code>{{radCreateDestroy}}</code>:
</p>
<rx-input>
  <select
    rx-select
    id="selTargetCreated"
    ng-if="radCreateDestroy === 'created'">
    <option>CREATED!</option>
  </select>
</rx-input>
```

### after
screen shot 2016-09-01 at 1 51 35 pm

File: src/elements/Forms/docs/examples/rxSelect.destroy.html [hunk 23]
```diff
@@ -0,0 +1,38 @@
+<form rx-form ng-controller="rxSelectDestroyCtrl">
+  <rx-form-section controlled-width>
+    <rx-field>
+      <rx-field-name></rx-field-name>
```
Comment (CITguy) [inline src/elements/Forms/docs/examples/rxSelect.destroy.html]: `<rx-field-name>` is unnecessary. Please remove.

File: src/elements/Forms/docs/forms.docs.html [hunk 24]
```diff
@@ -132,18 +132,157 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
...
+  <p>
+    <select id="plainSelDisabled" disabled>
+      <option value="">Disabled HTML Select Option</option>
+      <option value="first">First Option</option>
+      <option value="second">Second Option</option>
+      <option value="third">Third Option</option>
+      <option value="fourth">Fourth Option</option>
+    </select>
+  </p>
+  <p>
```
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: This select is not getting its value set correctly. 

This PR | Desired
----- | -----
screen shot 2016-09-01 at 2 13 12 pm | screen shot 2016-09-01 at 2 12 17 pm
Third select should be "Non Default Starting Option" |  



This can be fixed by moving the `ng-controller` attribute from `<rx-debug>` to this `<p>`

```html
<rx-debug>
  ...
  <p ng-controller="rxSelectCtrl">
    ...
  </p>
</rx-debug>
```

File: utils/rx-page-objects/test/rxSelect.midway.js [hunk 25]
```diff
@@ -55,24 +55,24 @@ describe('rxSelect', function () {
...
             selectedText: 'Disabled HTML Select Option'
         }));
 
         describe('Valid Enabled Non-Default Starting Value', encore.exercise.rxSelect({
-            instance: encore.rxSelect.initialize($('#plainHtmlSecondSelected')),
+            instance: encore.rxSelect.initialize($('#plainSelSecondSelected')),
             disabled: false,
             valid: true,
-            selectedText: 'Non Default Starting Option'
+            selectedText: 'Starts on Second Option'
```
Comment (CITguy) [inline utils/rx-page-objects/test/rxSelect.midway.js]: Please revert this change. The `#plainSelSecondSelected` should be "Non Default Starting Option".

File: src/elements/Forms/docs/forms.docs.html (hunk 24 above)
Comment (manojdas2212) [inline src/elements/Forms/docs/forms.docs.html]: @CITguy i did the same as you expected still i am getting this issue , i am not sure why "Non Default Starting Option" is not getting selected

screen shot 2016-09-02 at 4 18 13 pm

Comment (manojdas2212) [inline src/elements/Forms/docs/forms.docs.html]: screen shot 2016-09-02 at 5 42 04 pm

Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: You moved the `ng-controller` attribute from `<rx-debug>` to `<p>`, correct? I did that and had no issues, locally. 

I think the problem is that `<rx-debug>` is messing with the controller scope, so it's not seeing `htmlSelectAlternativeValue`.

File: src/elements/Forms/docs/forms.docs.html [hunk 26]
```diff
@@ -132,18 +132,159 @@ <h2 class="clear"><rx-permalink>Directives</rx-permalink></h2>
...
+          <option value="na">Disabled by 'disabled' attribute</option>
+        </select>
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+<rx-debug>
+  <p ng-controller="rxSelectCtrl">
+  <h3>Plain HTML Select Boxes (for comparison)</h3>
```
Comment (manojdas2212) [inline src/elements/Forms/docs/forms.docs.html]: @CITguy Could you please check it here "<p ng-controller="rxSelectCtrl">"

File: src/elements/Forms/docs/forms.docs.html (hunk 24 above)
Comment (CITguy) [inline src/elements/Forms/docs/forms.docs.html]: I see what's going on. You have to put the `ng-controller` on the `<p>` containing the dropdown in question, not wrap all of the existing paragraphs in a new `<p>`.

**Comment by manojdas2212** [comment]
@CITguy made changes please take a look

**Comment by CITguy** [comment]
Dev LGTM

**Comment by parlarjb** [comment]
Dev LGTM

**Comment by ssalinas24** [comment]
Design LGTM
